File: typescript/agent.ts
/**
 * Order coordinator agent — TypeScript example.
 *
 * Coordinates inventory reservation, payment, and shipping in sequence.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "order-coordinator-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const orderId = payload.order_id ?? "unknown";
  const items: any[] = payload.items ?? [];
  const customer = payload.customer_id ?? "unknown";
  const address = payload.shipping_address ?? "unknown";

  // Step 1: Inventory
  console.log(`  [1/3] Reserving inventory for ${items.length} item(s)...`);
  await new Promise((r) => setTimeout(r, 1000));

  // Step 2: Payment
  console.log(`  [2/3] Charging customer ${customer}...`);
  await new Promise((r) => setTimeout(r, 1000));

  // Step 3: Shipping
  console.log(`  [3/3] Creating shipment to ${address}...`);
  await new Promise((r) => setTimeout(r, 1000));

  const result = {
    action: "complete",
    order_id: orderId,
    inventory_reserved: true,
    payment_captured: true,
    tracking_number: "SHIP-44321",
    completed_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "order-coordinator-demo" });
  console.log(`  Order ${orderId} completed. Tracking: SHIP-44321`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;
    if (intentId && ["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: java/Agent.java
/*
 * Order coordinator agent — Java example.
 *
 * Fetches an intent by ID, coordinates inventory + payment + shipping,
 * and resumes with result.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   javac -cp axme-sdk.jar Agent.java
 *   java -cp .:axme-sdk.jar Agent <intent_id>
 */

import dev.axme.sdk.AxmeClient;
import dev.axme.sdk.AxmeClientConfig;
import dev.axme.sdk.RequestOptions;
import java.time.Instant;
import java.util.List;
import java.util.Map;

public class Agent {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Agent <intent_id>");
            System.exit(1);
        }

        String apiKey = System.getenv("AXME_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("Error: AXME_API_KEY not set.");
            System.exit(1);
        }

        String intentId = args[0];
        var client = new AxmeClient(AxmeClientConfig.forCloud(apiKey));

        System.out.println("Processing intent: " + intentId);

        var intentData = client.getIntent(intentId, new RequestOptions());
        @SuppressWarnings("unchecked")
        var intent = (Map<String, Object>) intentData.getOrDefault("intent", intentData);
        @SuppressWarnings("unchecked")
        var payload = (Map<String, Object>) intent.getOrDefault("payload", Map.of());
        if (payload.containsKey("parent_payload")) {
            @SuppressWarnings("unchecked")
            var pp = (Map<String, Object>) payload.get("parent_payload");
            payload = pp;
        }

        String orderId = (String) payload.getOrDefault("order_id", "unknown");
        @SuppressWarnings("unchecked")
        var items = (List<?>) payload.getOrDefault("items", List.of());
        String customer = (String) payload.getOrDefault("customer_id", "unknown");
        String address = (String) payload.getOrDefault("shipping_address", "unknown");

        // Step 1: Inventory
        System.out.println("  [1/3] Reserving inventory for " + items.size() + " item(s)...");
        Thread.sleep(1000);

        // Step 2: Payment
        System.out.println("  [2/3] Charging customer " + customer + "...");
        Thread.sleep(1000);

        // Step 3: Shipping
        System.out.println("  [3/3] Creating shipment to " + address + "...");
        Thread.sleep(1000);

        var result = Map.<String, Object>of(
            "action", "complete",
            "order_id", orderId,
            "inventory_reserved", true,
            "payment_captured", true,
            "tracking_number", "SHIP-44321",
            "completed_at", Instant.now().toString()
        );

        client.resumeIntent(intentId, result, new RequestOptions());
        System.out.println("  Order " + orderId + " completed. Tracking: SHIP-44321");
    }
}
